dashboard/modules/job/job_manager.py

```diff
@@ -224,7 +223,14 @@ async def run(
                     self._status_client.put_status(self._job_id,
                                                    JobStatus.SUCCEEDED)
                 else:
-                    log_tail = self._log_client.tail_logs(self._job_id)
+                    log_tail_iter = self._log_client.tail_logs(self._job_id)
+                    log_tail_deque = deque(maxlen=self.NUM_LOG_LINES_ON_ERROR)
+                    for line in log_tail_iter:
+                        if line is None:
+                            break
+                        else:
+                            log_tail_deque.append(line)
+                    log_tail = "".join(log_tail_deque)
```

Comment:
why do we need this deque behavior? I think this worked as-is before, no?

Reply:
we should just have a separate method on `tail_logs` instead of overloading the iterator (call it `get_last_n_log_lines` or something)

Reply:
with current implementation it will return an iterator instead of string and by default shows everything. This snippet just keep a fixed size deque and only keep the last NUM_LOG_LINES_ON_ERROR lines, default of 10